ouroboros-consensus-shelley-test/src/Test/ThreadNet/Infra/Shelley.hs

```diff
@@ -268,3 +307,195 @@ mkProtocolRealTPraos genesis CoreNode { cnDelegateKey, cnVRF, cnKES, cnOCert } =
+        replicateM (fromIntegral n) $
+        genCoreNode initialKESPeriod
+
+    numSlots     <-
+      if rotateKESKeys
+        then pure . NumSlots $ 30
+        else arbitrary
+
+    nodeRestarts <-
+      if rotateKESKeys
+        then ocertNodeRestarts nodeJoinPlan numSlots genesisConfig coreNodes (0,0)
+        else pure noRestarts
+
+    let
+      tc = TestConfig
+        { nodeTopology
+        , numCoreNodes
+        , numSlots
+        , initSeed
+        }
+
+    pure $ ShelleyTestConfig
+      { stcTestConfig = tc
+      , stcGenesis = genesisConfig
+      , stcCoreNodes = coreNodes
+      , stcNodeRestarts = nodeRestarts
+      }
+
+shrinkShelleyTestConfig :: ShelleyTestConfig -> [ShelleyTestConfig]
+shrinkShelleyTestConfig _ = [] -- TODO
+
+{-------------------------------------------------------------------------------
+  Node restarts for operational certificates
+-------------------------------------------------------------------------------}
+
+-- | Generate a node restart schedule which cycles nodes in order to replace
+-- their operational certificates after the maximum number of KES periods.
+--
+ocertNodeRestarts
+  :: NodeJoinPlan
+  -> NumSlots         -- ^ Duration of the test. No replacements will be scheduled
+                      --   after this number of slots.
+  -> ShelleyGenesis c -- ^ Shelley genesis configuration. Needed to extract the
+                      --   slots per KES period and the maximum number of KES
+                      --   evolutions.
+  -> [CoreNode c]     -- ^ Core nodes, including all of their keys.
+  -> (Word64, Word64) -- ^ Permitted (eagerness, laziness) in replacing operational
+                      --   certificates - that is, nodes may replace them up to
+                      --   `eagerness` KES periods early, and up to `laziness`
+                      --   KES periods late.
+  -> Gen NodeRestarts
+ocertNodeRestarts
+  (NodeJoinPlan _m)
+  (NumSlots t)
+  ShelleyGenesis {sgSlotsPerKESPeriod, sgMaxKESEvolutions}
+  coreNodes
+  (eagerness, laziness)
+    | t < ocertRegenCadence = pure noRestarts
+    | otherwise =
+      fmap NodeRestarts $ do
+        slotCNPairs <- fmap join $ forM (zip coreNodes [0..]) $ \(cn, coreNodeIdx) -> do
+          offsets <- Gen.infiniteListOf $ Gen.choose (- eagerness, laziness)
+          let SL.KESPeriod firstKESPeriod = SL.ocertKESPeriod $ cnOCert cn
+              periods =
+                unfoldr
+                  ( \(startPeriod, ix) -> let
+                      np = fromIntegral startPeriod + sgMaxKESEvolutions + offsets !! ix
+                      in Just $
+                    ( fromIntegral np
+                    , (fromIntegral np, ix + 1)
+                    )
+                  )
+                  (firstKESPeriod, 0)
+          pure
+            [ (SlotNo slot, CoreNodeId coreNodeIdx)
+              | slot <- takeWhile (< t) ((kesPeriodStartSlots !!) <$> periods)
+            ]
+        pure
```

Comment:
It's hard (at least for me) to see with which lines (415-418) these align. Extracting the code in the `forM` into a local helper will make this easier to read for me :slightly_smiling_face: